### PR TITLE
Emulate HTMLTrackElement to enable load event.

### PR DESCRIPTION
**Pull request by chemoish**
Figure out when track elements are loaded so we can manipulate cues without having to do setTimeout hacks.

Addresses https://github.com/videojs/video.js/issues/2784.

Also addresses (maybe) https://github.com/videojs/video.js/issues/2799.

### CODE REVIEW

File: src/js/control-bar/text-track-controls/chapters-button.js [hunk 1]
```diff
@@ -77,21 +77,13 @@ class ChaptersButton extends TextTrackButton {
     let chaptersTrack;
     let items = this.items = [];
 
-    for (let i = 0, l = tracks.length; i < l; i++) {
+    for (let i = 0, length = tracks.length; i < length; i++) {
       let track = tracks[i];
+
       if (track['kind'] === this.kind_) {
```
Comment (chemoish) [inline src/js/control-bar/text-track-controls/chapters-button.js]: Followed https://github.com/videojs/video.js/pull/1949

File: src/js/control-bar/text-track-controls/chapters-button.js [hunk 2]
```diff
@@ -105,7 +97,19 @@ class ChaptersButton extends TextTrackButton {
       }));
     }
 
-    if (chaptersTrack) {
+    if (chaptersTrack && !chaptersTrack.cues) {
```
Comment (chemoish) [inline src/js/control-bar/text-track-controls/chapters-button.js]: This seems questionable for me when determining when to add event listeners.

`chaptersTrack.cues` === null (native, null on instantiation—0 after setup)
`chaptersTrack.cues` === object (emulated)

File: src/js/tech/html5.js [hunk 3]
```diff
@@ -728,32 +728,35 @@ class Html5 extends Tech {
       return super.addRemoteTextTrack(options);
     }
 
-    var track = document.createElement('track');
+    var htmlTrackElement = document.createElement('track');
```
Comment (chemoish) [inline src/js/tech/html5.js]: There is a mixed bag between track, trackElement, htmlTrackElement, textTrackEls, etc (Tried my best to follow existing convention with adding a bit more specificity).

File: src/js/tech/html5.js [hunk 4]
```diff
@@ -767,7 +770,11 @@ class Html5 extends Tech {
       return super.removeRemoteTextTrack(track);
     }
 
-    var tracks, i;
+    let tracks, i;
+
+    // TODO: track can be html text element due to addRemoteTextTrack—consider refactoring
```
Comment (chemoish) [inline src/js/tech/html5.js]: Forgot to attach issue, but same thing as html tech.

File: src/js/tech/html5.js [hunk 5]
```diff
@@ -767,7 +770,11 @@ class Html5 extends Tech {
...
+    let trackElement = this.remoteTextTrackEls().getTrackElementByTrack_(track);
```
Comment (chemoish) [inline src/js/tech/html5.js]: This confused me quite a bit (also due to the removeTrack_ bug), but `addRemoteTextTrack` returns an `HTMLTrackElement` or `{track: TextTrack}`—neither of which are `TextTrack` capable of being removed.

**Comment by pam** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: d318bcc2d401d2744a7454e9c00bcd47435c0b4f
Build details: https://travis-ci.org/pam/video.js/builds/90449472

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: src/js/tracks/html-track-element-list.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/**
+ * @file html-track-element.js
+ */
+
+import * as browser from '../utils/browser.js';
+
+class HtmlTrackElementList {
```
Comment (chemoish) [inline src/js/tracks/html-track-element-list.js]: Followed `TextTrackList` even if this wasn't shimming anything.

File: src/js/tracks/text-track.js [hunk 7]
```diff
@@ -233,22 +234,23 @@ TextTrack.prototype.removeCue = function(removeCue) {
...
 
   parser['oncue'] = function(cue) {
     track.addCue(cue);
   };
+
   parser['onparsingerror'] = function(error) {
     log.error(error);
   };
 
+  parser['onflush'] = () => {
```
Comment (chemoish) [inline src/js/tracks/text-track.js]: es6 unnecessary.

File: src/js/tracks/text-track.js [hunk 8]
```diff
@@ -269,7 +271,15 @@ var loadTrack = function(src, track) {
     }
 
     track.loaded_ = true;
-    parseCues(responseBody, track);
+
+    if (typeof window['WebVTT'] !== 'function') {
+      //try again a bit later
+      return window.setTimeout(function() {
+        parseCues(responseBody, track);
+      }, 100);
```
Comment (chemoish) [inline src/js/tracks/text-track.js]: arbitrary increase—seemed to work better on my local environment.

File: src/js/tracks/text-track.js [hunk 9]
```diff
@@ -269,7 +271,15 @@ var loadTrack = function(src, track) {
     }
 
     track.loaded_ = true;
-    parseCues(responseBody, track);
+
+    if (typeof window['WebVTT'] !== 'function') {
+      //try again a bit later
+      return window.setTimeout(function() {
```
Comment (chemoish) [inline src/js/tracks/text-track.js]: return is now unnecessary.

**Comment by gkatsev** [comment]
Awesome, I'll take a look tomorrow.

**Comment by chemoish** [comment]
Will need help:

1. browser compatibility
1. testing strategy

Works (OSX 10.10.5):
- Chrome 46 (mp4, mpeg-dash)
- Safari 8.0.8
- Firefox 40
- IE 11 Windows 8.1 (virtualbox)

`this.remoteTextTracks()` and `this.remoteTextTrackEls()` work as expected.

File: src/js/tracks/text-track.js [hunk 10]
```diff
@@ -269,7 +271,15 @@ var loadTrack = function(src, track) {
     }
 
     track.loaded_ = true;
-    parseCues(responseBody, track);
+
+    if (typeof window['WebVTT'] !== 'function') {
```
Comment (chemoish) [inline src/js/tracks/text-track.js]: add comments if this isn't an issue in any environment beyond local.

File: src/js/control-bar/text-track-controls/chapters-button.js (hunk 2 above)
Comment (gkatsev) [inline src/js/control-bar/text-track-controls/chapters-button.js]: Emulated captions should be returning `null` until the track loads. https://github.com/videojs/video.js/blob/0e1c424e48019b002ba542aaceb211b959f511a0/src/js/tracks/text-track.js#L126-L135

File: src/js/control-bar/text-track-controls/chapters-button.js [hunk 11]
```diff
@@ -105,7 +97,19 @@ class ChaptersButton extends TextTrackButton {
...
+      chaptersTrack['mode'] = 'hidden';
+
+      let remoteTextTrackEl = this.player_.remoteTextTrackEls().getTrackElementByTrack_(chaptersTrack);
+
+      if (remoteTextTrackEl) {
+        remoteTextTrackEl.addEventListener('load', (event) => {
```
Comment (gkatsev) [inline src/js/control-bar/text-track-controls/chapters-button.js]: the curly braces around `this.update()` aren't necessary in this case.

File: src/js/tech/html5.js [hunk 12]
```diff
@@ -728,32 +728,35 @@ class Html5 extends Tech {
       return super.addRemoteTextTrack(options);
     }
 
-    var track = document.createElement('track');
+    var htmlTrackElement = document.createElement('track');
 
     if (options['kind']) {
-      track['kind'] = options['kind'];
+      htmlTrackElement['kind'] = options['kind'];
```
Comment (gkatsev) [inline src/js/tech/html5.js]: While you're already here, want to change all the bracket notation to dot notation?
`htmlTrackEleemnt['kind'] --> htmlTrackElement.kind`

File: src/js/tech/html5.js (hunk 5 above)
Comment (gkatsev) [inline src/js/tech/html5.js]: Hopefully, with this update, this should only ever be an official `HTMLTrackElement` or our emulated implementation and not `{track: TextTrack}`

File: src/js/tech/tech.js [hunk 13]
```diff
@@ -371,6 +373,17 @@ class Tech extends Component {
   }
 
   /**
+   * Get remote htmltrackelements
+   *
+   * @returns {HTMLTrackElementList}
+   * @method remoteTextTrackEls
+   */
+  remoteTextTrackEls() {
+    this.remoteTextTrackEls_ = this.remoteTextTrackEls_ || new HTMLTrackElementList();
```
Comment (gkatsev) [inline src/js/tech/tech.js]: We'll also probably want to go through the same shenanigans we've gone through with the text track list to have it work across tech changes. See https://github.com/videojs/video.js/pull/2391
This can, and probably should, be tackled only once we have everything else working and could probably event go out as a separate PR.

File: src/js/tech/tech.js [hunk 14]
```diff
@@ -389,19 +402,24 @@ class Tech extends Component {
...
+
+    this.remoteTextTrackEls().addTrackElement_(htmlTrackElement);
     this.remoteTextTracks().addTrack_(track);
-    return {
-      track: track
-    };
+
+    // TODO: need to verify that we should be returning htmlTrackElement
+    // https://github.com/videojs/video.js/issues/2799
+    return htmlTrackElement;
```
Comment (gkatsev) [inline src/js/tech/tech.js]: Yes, I believe we should be returning the htmlTrackElement here.

File: src/js/tech/html5.js (hunk 3 above)
Comment (gkatsev) [inline src/js/tech/html5.js]: also, can you change the `var` to a `let`?

File: src/js/tracks/html-track-element-list.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  constructor(trackElements = []) {
+    this.trackElements_ = [];
+
+    Object.defineProperty(this, 'length', {
```
Comment (gkatsev) [inline src/js/tracks/html-track-element-list.js]: unfortunately, for IE8, we need to use a custom object because IE8 only lets use `defineProperty` on dom elements.
See https://github.com/videojs/video.js/blob/master/src/js/tracks/text-track.js#L43-L50 for a template.

File: src/js/tracks/html-track-element.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    this.srclang = track.srclang;
+    this.label = track.label;
+    this.default = track.default;
+
+    const NONE = 0;
+    const LOADING = 1;
+    const LOADED = 2;
+    const ERROR = 3;
+
+    Object.defineProperty(this, 'readyState', {
```
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: Same as above, see https://github.com/videojs/video.js/blob/master/src/js/tracks/text-track.js#L43-L50

File: src/js/tracks/html-track-element.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+      // TODO: do stuff
+    }
+
+    this.kind = track.kind;
+    this.src = track.src;
+    this.srclang = track.srclang;
+    this.label = track.label;
+    this.default = track.default;
+
+    const NONE = 0;
```
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: can we export these on the element maybe?

File: src/js/tracks/text-track.js [hunk 18]
```diff
@@ -195,7 +195,8 @@ TextTrack.prototype.constructor = TextTrack;
  * cuechange - One or more cues in the track have become active or stopped being active.
  */
 TextTrack.prototype.allowedEvents_ = {
-  'cuechange': 'cuechange'
+  'cuechange': 'cuechange',
+  'loadeddata': 'loadeddata'
```
Comment (gkatsev) [inline src/js/tracks/text-track.js]: we actually don't want to add `loadeddata` here. We can still fire it but it should be a "private" event, like `modechange`.

File: src/js/tracks/text-track.js (hunk 10 above)
Comment (gkatsev) [inline src/js/tracks/text-track.js]: This is for the `alt/video.novtt.js` build.

File: src/js/tracks/text-track.js (hunk 9 above)
Comment (gkatsev) [inline src/js/tracks/text-track.js]: Makes sense.

File: src/js/tracks/text-track.js (hunk 8 above)
Comment (gkatsev) [inline src/js/tracks/text-track.js]: Sounds fine.

File: src/js/tracks/html-track-element.js [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    Object.defineProperty(this, 'track', {
+      get() {
+        return textTrack;
+      }
+    });
+
+    readyState = NONE;
+    textTrack = track;
+
+    this.allowedEvents_ = {
```
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: should move to the prototype

**Comment by gkatsev** [comment]
Thanks, this is definitely the right track, if you will ;)

File: src/js/tech/html5.js (hunk 3 above)
Comment (chemoish) [inline src/js/tech/html5.js]: I misread the `let` documentation… Is there ever a reason not to use `let`?

File: src/js/control-bar/text-track-controls/chapters-button.js (hunk 11 above)
Comment (chemoish) [inline src/js/control-bar/text-track-controls/chapters-button.js]: sounds good, my code style is much different—better to keep convention.

File: src/js/tracks/html-track-element.js (hunk 17 above)
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: Move the constants outside of `HTMLTrackElement` and export them?—Is this what you mean?

File: src/js/tracks/html-track-element.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+      readyState = LOADED;
+
+      trackElement.trigger({
+        type: 'load',
+        target: trackElement
+      });
+    });
+
+    if (browser.IS_IE8) {
+      return this;
```
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: fix ^

File: src/js/tracks/html-track-element.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+class HTMLTrackElement extends EventTarget {
+  constructor(track) {
+    super();
+
+    var readyState,
+        textTrack,
+        trackElement = this;
+
+    if (browser.IS_IE8) {
+      // TODO: do stuff
```
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: fix ^

File: src/js/tracks/html-track-element-list.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        return this.trackElements_.length;
+      }
+    });
+
+    for (let i = 0, length = trackElements.length; i < length; i++) {
+      this.addTrackElement_(trackElements[i]);
+    }
+
+    if (browser.IS_IE8) {
+      return this;
```
Comment (chemoish) [inline src/js/tracks/html-track-element-list.js]: fix ^

File: src/js/tracks/html-track-element.js [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ *   readonly attribute TextTrack track;
+ * };
+ */
+
+class HTMLTrackElement extends EventTarget {
+  constructor(track) {
+    super();
+
+    var readyState,
```
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: let

File: src/js/tracks/html-track-element.js [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+    });
+
+    Object.defineProperty(this, 'track', {
+      get() {
+        return textTrack;
+      }
+    });
+
+    readyState = NONE;
+    textTrack = track;
```
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: Refactor out `textTrack`

File: src/js/tech/html5.js (hunk 3 above)
Comment (gkatsev) [inline src/js/tech/html5.js]: Not really.

Comment (misteroneill) [inline src/js/tech/html5.js]: Yeah, I don't see a reason to ever use anything but `let` and `const`.

File: src/js/tracks/html-track-element-list.js [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+class HtmlTrackElementList {
+  constructor(trackElements = []) {
+    let list = this;
+
+    if (browser.IS_IE8) {
+      list = document.createElement('custom');
+
+      for (let prop in HtmlTrackElementList.prototype) {
+        list[prop] = HtmlTrackElementList.prototype[prop];
```
Comment (gkatsev) [inline src/js/tracks/html-track-element-list.js]: Just remembered, we should also make sure that we don't copy the `constructor` property because IE8 will complain.

**Comment by pam** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 54b35f881b1d1968f4c6e56ee65aac362145fcc7
Build details: https://travis-ci.org/pam/video.js/builds/91732554

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: src/js/tracks/html-track-element.js [hunk 26]
```diff
@@ -0,0 +1,86 @@
+import * as browser from '../utils/browser.js';
+import document from 'global/document';
+import EventTarget from '../event-target';
+
+export const NONE = 0;
```
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: We've generally been putting all exports at the bottom of a file.
Though, I meant more like:
```js
HTMLTrackElement.NONE = 0;
HTMLTrackElement.LOADING = 1;
// ...
```
Sorry I wasn't more clear earlier.

File: src/js/tracks/text-track-list.js [hunk 27]
```diff
@@ -89,15 +89,22 @@ TextTrackList.prototype.addTrack_ = function(track) {
...
-  let result = null;
   let track;
 
   for (let i = 0, l = this.length; i < l; i++) {
-    track = this[i];
-    if (track === rtrack) {
-      this.tracks_.splice(i, 1);
-      break;
+    if (this[i] !== rtrack) {
+      continue;
```
Comment (gkatsev) [inline src/js/tracks/text-track-list.js]: Can we keep it so that it only does the splice and break if they match instead of continuing? It'll be more inline with how everything else is written.

File: src/js/tracks/text-track-list.js [hunk 28]
```diff
@@ -89,15 +89,22 @@ TextTrackList.prototype.addTrack_ = function(track) {
...
     }
+
+    track = this[i];
+
+    this.tracks_.splice(i, 1);
+
+    break;
+  }
+
+  if (track) {
```
Comment (gkatsev) [inline src/js/tracks/text-track-list.js]: Should this check for `!track`? Otherwise we'll never trigger `removetrack` if we actually found and removed a track?

File: src/js/tracks/html-track-element.js (hunk 26 above)
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: These consts could still be used/exported, in which case you could do
```js
const NONE = 0;
 
// ... use NONE inside the class

HTMLTrackElement.NONE  = NONE;

// if we still want to export it
export NONE;
```

File: src/js/tracks/html-track-element-list.js (hunk 25 above)
Comment (chemoish) [inline src/js/tracks/html-track-element-list.js]: `text-track`, `text-track-list`, etc. are all doing the same thing.

Will the final es6 implementation utilize constructor() differently than TextTrack.prototype.constructor?

File: src/js/tracks/text-track-list.js (hunk 27 above)
Comment (chemoish) [inline src/js/tracks/text-track-list.js]: Yup—habit of minimizing indentation and getting out early..

File: src/js/tracks/text-track-list.js (hunk 28 above)
Comment (chemoish) [inline src/js/tracks/text-track-list.js]: Good catch, pretty much inverted the bug ><

**Comment by chemoish** [comment]
Very troubling that the branch no longer worked like expected. https://github.com/videojs/video.js/pull/2804#issuecomment-155650936

Retested all the same environments with new code. Works…

**Comment by pam** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 0212fcb7fb851a9e1506f529957af6025f610306
Build details: https://travis-ci.org/pam/video.js/builds/91748257

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: src/js/tracks/html-track-element-list.js (hunk 25 above)
Comment (gkatsev) [inline src/js/tracks/html-track-element-list.js]: Looks like not all got updated by this PR: https://github.com/videojs/video.js/pull/2565/files

Not sure what you mean by the second part but babel will always compile it down to prototype stuff. Though, the prototype still exists in full classes as well, I believe.

File: src/js/tracks/html-track-element.js [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+    trackElement.default = track.default;
+
+    Object.defineProperty(trackElement, 'readyState', {
+      get() {
+        return readyState;
+      }
+    });
+
+    Object.defineProperty(trackElement, 'track', {
+      get() {
```
Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: Does not having a `set()` prevent it from getting set, I assume?

**Comment by gkatsev** [comment]
Good stuff and all the tests pass currently. I'll make a comment about unit testing tomorrow but otherwise, you can take a look at how `text-track.test.js` is or some of the others.
Also, I'll help out testing locally.

File: src/js/tracks/html-track-element.js (hunk 29 above)
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: Yes. `set()` defaults to undefined. I just tried to enforce the readonly interface for `TextTrack`.

Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: Yep, :100:% agree. Just wasn't certain. :+1:

Comment (misteroneill) [inline src/js/tracks/html-track-element.js]: Might be worth taking into consideration the performance implications of `defineProperty`. I was reading [this article](https://www.nczonline.net/blog/2015/11/performance-implication-object-defineproperty/) by Nicholas Zakas yesterday on that topic. [More here](https://code.google.com/p/v8/issues/detail?id=3649).

Comment (gkatsev) [inline src/js/tracks/html-track-element.js]: It's slow but happens in the constructor, so, not *that* bad but also, we don't really have a choice since we want read-only properties and also we want an API that matches the spec, so, getters and setters are the only way.

Comment (chemoish) [inline src/js/tracks/html-track-element.js]: Good to know though.

File: src/js/tracks/html-track-element-list.js (hunk 25 above)
Comment (chemoish) [inline src/js/tracks/html-track-element-list.js]: PR makes it more clear.

**Comment by pam** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: cb65a1241b2081d1de2f16d2abc23f4ab12813bc
Build details: https://travis-ci.org/pam/video.js/builds/91940747

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by pam** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 97c0138422089415cf8775b4d86f64263510f671
Build details: https://travis-ci.org/pam/video.js/builds/91940718

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: src/js/tracks/html-track-element.js (hunk 29 above)
Comment (chemoish) [inline src/js/tracks/html-track-element.js]: https://github.com/videojs/video.js/blob/master/src/js/tracks/text-track-cue-list.js#L53
https://github.com/videojs/video.js/blob/master/src/js/tracks/text-track-list.js#L75

File: test/unit/tracks/html-track-element.test.js [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+  window.throws(
+    function() {
+      new HTMLTrackElement();
+    },
+    new Error('A tech was not provided.'),
+    'a tech is required for html track element'
+  );
+});
+
+test('can create a html track element with various properties', function() {
```
Comment (chemoish) [inline test/unit/tracks/html-track-element.test.js]: not sure about test requirements

File: test/unit/tracks/html-track-element.test.js [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+  equal(htmlTrackElement.default, undefined, 'we have a default');
+  equal(htmlTrackElement.kind, kind, 'we have a kind');
+  equal(htmlTrackElement.label, label, 'we have a label');
+  equal(htmlTrackElement.readyState, 0, 'we have a readyState');
+  equal(htmlTrackElement.src, src, 'we have a src');
+  equal(htmlTrackElement.srclang, language, 'we have a srclang');
+  equal(htmlTrackElement.track instanceof TextTrack, true, 'we have a track');
+});
+
+test('defaults when items not provided', function() {
```
Comment (chemoish) [inline test/unit/tracks/html-track-element.test.js]: not sure about test requirements

**Comment by pam** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: eb538cb68f9f7722aae14ed147025f8fdf9bd7bc
Build details: https://travis-ci.org/pam/video.js/builds/91963902

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: test/unit/tracks/html-track-element.test.js [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+  equal(htmlTrackElement.kind, 'subtitles', 'we have a kind');
+  equal(htmlTrackElement.label, '', 'we have a label');
+  equal(htmlTrackElement.readyState, 0, 'we have a readyState');
+  equal(htmlTrackElement.src, undefined, 'we have a src');
+  equal(htmlTrackElement.srclang, '', 'we have a srclang');
+  equal(htmlTrackElement.track instanceof TextTrack, true, 'we have a track');
+});
+
+test('fires loadeddata when track cues become populated', function() {
+  let player = TestHelpers.makePlayer(),
```
Comment (chemoish) [inline test/unit/tracks/html-track-element.test.js]: unnecessary

**Comment by pam** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 665608304c28e35fdaf9aeab6ce03ab9abd632c1
Build details: https://travis-ci.org/pam/video.js/builds/92333120

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by pam** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 6bbe8fb12117db4e445227a4793a290513f6343a
Build details: https://travis-ci.org/pam/video.js/builds/92334437

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by raytiley** [comment]
So I built this and tested it out and using the js api is working well now. I still can't get a chapter button to appear if using a `track` element:

works:

```html
<video id='player' controls=true class='video-js vjs-default-skin vjs-big-play-centered'>
                <source src="http://example.com/video.mp4" type="video/mp4" />
                Your browser does not support html5 video.
            </video>
```

```js
$('document').ready(function () {
                var player= videojs('player');
                var track = player.addRemoteTextTrack({
                    kind: 'chapters',
                    language: 'en',
                    label: 'English',
                    src: 'http://example.com/video.vtt'
                });
            });

```

doesn't work

```html
<video id='player' controls=true class='video-js vjs-default-skin vjs-big-play-centered'>
                <source src="http://example.com/video.mp4" type="video/mp4" />
                <track kind="chapters" src="http://example.com/video.vtt" srclang="en" label="Chapters" default />
                Your browser does not support html5 video.
            </video>
```

```js
$('document').ready(function () {
                var player= videojs('player');
            });
```

Am I doing something wrong? Thanks for the help.

**Comment by gkatsev** [comment]
The track element issue is probably related to some of our other issues: https://github.com/videojs/video.js/issues/2689, https://github.com/videojs/video.js/issues/2542, https://github.com/videojs/video.js/pull/2835

**Comment by vhmth** [comment]
@raytiley I'm not even able to get the programmatic approach to work.

*relevant markup*

```
<video controls poster="<%= videoPosterURL %>" class="video-js vjs-default-skin vjs-big-play-centered vjs-fluid">
  <source src="<%= videoURL %>" type="video/mp4" />
</video>
```

*relevant js*

```
var player = videojs(this.$video[0], {}, function () {
  var playTracked = false;
  // we track the first video play only
  this.on('play', function () {
    // we can't simply 'off' the play event since other UI elements depend on it
    if (!playTracked) {
      mixpanel.track(mixpanelEvents.FEEDBACK_VIDEO_PLAYED);
      playTracked = true;
    }
  });
});

player.addRemoteTextTrack({
  kind: 'chapters',
  language: 'en',
  src: '/assets/demo.vtt'
});
```

*demo.vtt*

```
WEBVTT

1
00:00:00 --> 00:00:22
Chapter 1

2
00:00:22 --> 00:00:30
Short chapter

3
00:00:30 --> 00:00:45
Finale
```

File: src/js/control-bar/text-track-controls/chapters-button.js [hunk 33]
```diff
@@ -105,7 +97,17 @@ class ChaptersButton extends TextTrackButton {
       }));
     }
 
-    if (chaptersTrack) {
+    if (chaptersTrack && chaptersTrack.cues == null) {
+      chaptersTrack['mode'] = 'hidden';
+
+      let remoteTextTrackEl = this.player_.remoteTextTrackEls().getTrackElementByTrack_(chaptersTrack);
```
Comment (gkatsev) [inline src/js/control-bar/text-track-controls/chapters-button.js]: looks like there an issue here tracks that were added in the html:
```html
  <video id="vid1" class="video-js vjs-default-skin" controls preload="auto" width="640" height="264"
      poster="http://video-js.zencoder.com/oceans-clip.png">
    <source src="oceans-clip.mp4" type='video/mp4'>
    <source src="http://video-js.zencoder.com/oceans-clip.mp4?fo23" type='video/mp4'>
    <track kind="chapters" src="../docs/examples/shared/example-captions.vtt" srclang="en" label="Subs"></track>
</video>
```
What happens here is that we get the chapters track but when we try to retrieve its associated track el, we get undefined so the chapters button never loads.
I think what we need to do is in [`proxyNativeTextTracks_`](https://github.com/videojs/video.js/blob/a601fbb83fdc442f30f09ca15ff55b04c3e7c91d/src/js/tech/html5.js#L254) to loop through any track elements and add them to the remoteTextTrackEls list. And then in [`addRemoteTextTrack`](https://github.com/videojs/video.js/blob/a601fbb83fdc442f30f09ca15ff55b04c3e7c91d/src/js/tech/html5.js#L760) we need to make sure the track el gets added to the list. We also want to make sure that `remoteRemoteTextTrack` removes the associated el.
`html-track-element-list` might need to get updated to support this but hopefully it'll just work.

**Comment by gkatsev** [comment]
There's one thing left to do, otherwise, this looks great.

File: test/unit/player.test.js [hunk 34]
```diff
@@ -835,6 +835,39 @@ test('createModal()', function() {
   strictEqual(player.children().indexOf(modal), -1, 'modal was removed from player\'s children');
 });
 
+test('should return correct remote text track values', function () {
```
Comment (chemoish) [inline test/unit/player.test.js]: Not sure where this should go or how this should be tested.

- Note: `techOrder: ['html5']`

---

`techFaker` doesn't initialize the `<track>` correctly, but fires the `player.addRemoteTextTrack` correctly. Open to recommendations.

**Comment by chemoish** [comment]
Tested `<track>`:

- IE 11 Win 8.1 (virtualbox)
- Chrome (latest)
- Safari (latest)
- FF (latest)

Tested `addRemoteTextTrack`:

- IE 11 Win 8.1 (virtualbox)
- Chrome (latest)
- Safari (latest)
- FF (latest)

All passed.

**Comment by chemoish** [comment]
Tests pass locally.

File: test/unit/player.test.js (hunk 34 above)
Comment (gkatsev) [inline test/unit/player.test.js]: I think it probably would be better to move this test into `test/tracks/tracks.test.js`

**Comment by gkatsev** [comment]
We also should add some tests for the new functionality of `addRemoteTextTrack` as well as tests for `remoteTextTracksEls()`

**Comment by chemoish** [comment]
I am down for better tests, but what specifically did you have in mind?

The only new functionality specified for `addRemoteTextTrack` was the addition of managing the `remoteTextTrackEls`.

- https://github.com/videojs/video.js/pull/2804/files?diff=unified#diff-5c5c01f73339e29a9ae3816bd7fd777d (should cover `remoteTextTrackEls()`)
- https://github.com/videojs/video.js/pull/2804/files?diff=unified#diff-ff7bdfd93378f289476b912a6a3e07afR838 (should cover `remoteTextTrackEls` instantiation; `addRemoteTextTrack`, `remoteRemoteTextTrack` verification).

`tech::addRemoteTextTrack` did unify the return of the htmlTrackElement, would that be beneficial to test?

**Comment by gkatsev** [comment]
Yep, basically just a simple test that verifies that when you call `addRemoteTextTrack` you can then see the element that was returned in the `remoteTextTrackEls` list.

**Comment by gkatsev** [comment]
Sweet, LGTM.

**Comment by gkatsev** [comment]
Merged! Thanks @chemoish. Also, thanks @OwenEdwards for the initial workup oh so long ago :)
